core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -299,30 +410,18 @@ class half_t {
 #ifdef __CUDA_ARCH__
     val += rhs.val;
 #else
-    val = __float2half(__half2float(val) + __half2float(rhs.val));
+    val     = __float2half(__half2float(val) + __half2float(rhs.val));
 #endif
     return *this;
   }
 
   KOKKOS_FUNCTION
-  volatile half_t& operator+=(half_t rhs) volatile {
-#ifdef __CUDA_ARCH__
-    // Cuda 10 supports __half volatile stores but not volatile arithmetic
-    // operands. Cast away volatile-ness of val for arithmetic but not for store
-    // location.
-    val = const_cast<impl_type&>(val) + rhs.val;
-#else
-    // Use non-volatile val_ref to suppress:
-    // "warning: implicit dereference will not access object of type ‘volatile
-    // __half’ in statement"
-    auto val_ref = const_cast<impl_type&>(val);
-    val_ref      = __float2half(__half2float(const_cast<impl_type&>(val)) +
-                           __half2float(rhs.val));
-#endif
+  volatile half_t operator+=(const volatile half_t rhs) volatile {
```

Comment:
I believe the argument should be `const volatile half_t &rhs` (reference, not value) if you want to actually pass it through as `volatile`. Or, if you're ok with the compiler making the coherent read and copy to a temporary in the function call, use the original signature, `volatile half_t& operator+=(half_t rhs) volatile`. I think the latter is preferable, since there's no need to pass the argument around as a reference if we're guaranteed to need a temporary copy anyway.